Add InkWell docs on transitions and ink splash clipping

shihaohong opened the pull request:
## Description

Added a section to the `InkWell` documentation highlighting that widget transitions within an InkWell widget may cause clipping or unpleasant rippling effects.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/5211

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

shihaohong committed:
Added InkWell docs on transitions causing clipping

shihaohong committed:
remove analysis benchmark file

shihaohong committed:
clarified header

HansMuller commented on packages/flutter/lib/src/material/ink_well.dart:
```diff
@@ -616,6 +616,13 @@ class _InkResponseState<T extends InkResponse> extends State<T> with AutomaticKe
 /// ancestor to the ink well). The [MaterialType.transparency] material
 /// kind can be used for this purpose.
 ///
+/// ### The ink splashes are unpleasantly animating with transitioning widgets!
```
I think this could be a little clearer.

The original issue made the problem pretty clear: [InkWell doesn't work well in AnimatedContainer](https://github.com/flutter/flutter/issues/5211).

Maybe this section could be called: InkWell's splashes don't track the size of an animated container

Start with paragraph that explains the symptoms of the problem (we want to the reader to be able to quickly confirm that this section is about the problem they're seeing). You could even include the example from the original issue.

This next paragraph is correct but probably a bit difficult to understand "transitions should not be placed inside Material" sounds draconian and we haven't tied Material and InkWell together here.  The overall point is that an InkWell's splashes will not change to conform to changes in the size of its underlying Material - where the splashes are actually rendered.  Typically the underlying Material comes from the MaterialApp and is not animated. We're saying that app's shouldn't create InkWells within Material widgets whose size is also animated. 

Once that's clear you could reiterate the bit about LayoutChangedNotification. However I don't think that will help most readers.

shihaohong replied:
Sounds good. I redid the explanation and title and decided to remove the LayoutChangedNotification paragraph, since it will likely not be helpful for our readers.

shihaohong replied:
After investigation, decided on an explanation as to why the ink effects to appear unusually. I also added a sample app to demonstrate the behavior.

shihaohong committed:
Update InkWell doc to explain clipping ink splashes

shihaohong committed:
Corrected description of nesting InkWell

shihaohong commented:
Had an offline discussion with @HansMuller, we had misunderstood the actual source of the clipping, so I'll be playing around InkWell and an animated child that changes size before I come up with an updated description of the issue

shihaohong committed:
Added sample clipping app

shihaohong committed:
Shorten and clarified inkwell clipping sample

HansMuller commented on packages/flutter/lib/src/material/ink_well.dart:
```diff
@@ -616,6 +616,48 @@ class _InkResponseState<T extends InkResponse> extends State<T> with AutomaticKe
+///
+/// An InkWell's splashes will not properly update to conform to changes in the
+/// size of its underlying [Material], where the splashes are rendered,
+/// mid-animation. Hence, you should avoid using InkWells within [Material]
```
mid-animation => changes during the animation

You can leave out "Hence, ".

HansMuller commented on packages/flutter/lib/src/material/ink_well.dart:
```diff
@@ -616,6 +616,48 @@ class _InkResponseState<T extends InkResponse> extends State<T> with AutomaticKe
 /// kind can be used for this purpose.
 ///
+/// ### The ink splashes don't track the size of an animated container
+/// When you wrap an animating widget with an InkWell, you may notice that ink
```
If the size of an InkWell's Material ancestor changes while the InkWell's splashes are expanding, you may notice that the splashes aren't clipped correctly. This can't be avoided.

HansMuller reviewed (approved):
LGTM with just a little more tweaking.

shihaohong committed:
Correct InkWell doc grammar

shihaohong committed:
Improve explanation of clipped InkWell

Incorporated code review feedback to clarify that clipping cannot be
avoided with growing animated containers. Further correct grammar.

shihaohong committed:
Fix code sample typo
